modules/extensions-support/src/main/java/org/mule/module/extension/internal/runtime/resolver/MetadataContextArgumentResolver.java

```diff
@@ -0,0 +1,79 @@
+
+import static org.mule.module.extension.internal.ExtensionProperties.ENCODING_PARAMETER_NAME;
+import static org.mule.module.extension.internal.ExtensionProperties.METADATA_CONTEXT;
+import static org.mule.module.extension.internal.ExtensionProperties.MIME_TYPE_PARAMETER_NAME;
+import org.mule.api.MuleMessage;
+import org.mule.extension.api.runtime.ContentType;
+import org.mule.extension.api.runtime.FixedContextMetadata;
+import org.mule.extension.api.runtime.ContentMetadata;
+import org.mule.extension.api.runtime.MutableContentMetadata;
+import org.mule.extension.api.runtime.OperationContext;
+import org.mule.module.extension.internal.ExtensionProperties;
+import org.mule.module.extension.internal.runtime.OperationContextAdapter;
+
+import org.apache.commons.lang.StringUtils;
+
+/**
+ * Returns the value of the {@link ExtensionProperties#METADATA_CONTEXT} variable,
+ * which is expected to have been previously set on the supplied {@link OperationContext}.
+ * <p>
+ * Notice that for this to work, the {@link OperationContext}
+ * has to be an instance of {@link OperationContextAdapter}
+ *
+ * @since 4.0
+ */
+public final class MetadataContextArgumentResolver implements ArgumentResolver<ContentMetadata>
+{
+
+    /**
+     * @param operationContext an {@link OperationContext}
+     * @return the value of {@link OperationContextAdapter#getVariable(String)} using the {@link ExtensionProperties#METADATA_CONTEXT}
+     * @throws ClassCastException       if {@code operationContext} is not an instance of {@link OperationContextAdapter}
+     * @throws IllegalArgumentException if the variable has not been set on the {@code operationContext}
+     */
+    @Override
+    public ContentMetadata resolve(OperationContext operationContext)
+    {
+        final OperationContextAdapter operationContextAdapter = (OperationContextAdapter) operationContext;
+        final ContentType currentContentType = getCurrentContentType(operationContextAdapter);
+        String encoding = nullableString(operationContext, ENCODING_PARAMETER_NAME);
+        String mimeType = nullableString(operationContext, MIME_TYPE_PARAMETER_NAME);
+
+        boolean contentTypeFixed = !StringUtils.isBlank(encoding) || !StringUtils.isBlank(mimeType);
+
+        ContentMetadata contentMetadata = contentTypeFixed
+                                          ? new FixedContextMetadata(currentContentType, new ContentType(encoding, mimeType))
+                                          : new MutableContentMetadata(currentContentType);
+
+        operationContextAdapter.setVariable(METADATA_CONTEXT, contentMetadata);
+        return contentMetadata;
+    }
+
+    private ContentType getCurrentContentType(OperationContextAdapter operationContext) {
+        MuleMessage message = operationContext.getEvent().getMessage();
+        return new ContentType(message.getEncoding(), message.getMimeType());
+    }
+
+    private String nullableString(OperationContext operationContext, String key)
```

Comment:
The idea of "nullable" argument is something that can be reused. It would be good to have a generic way to do that in the OperationContext or some place where it can be shared